docs/design.md

```diff
@@ -0,0 +1,258 @@
+    additionalProperties: false
+    description: Asserts that all of the nested rules match.
+    properties:
+      allOf:
+        $ref: '#/definitions/rules'
+    required:
+      - allOf
+    type: object
+  anyOfRule:
+    additionalProperties: false
+    description: Asserts that at least one of the nested rules matches.
+    properties:
+      anyOf:
+        $ref: '#/definitions/rules'
+    required:
+      - anyOf
+    type: object
+  matchRule:
+    additionalProperties: false
+    description: A rule for matching trust anchors.
+    properties:
+      match:
+        description: |
+          Defines the pattern to match trust anchors against. Can be an exact
+          string or a regular expression.
+        type: string
+    required:
+      - match
+    type: object
+  notRule:
+    additionalProperties: false
+    description: Inverts the matching behaviour of a rule.
+    properties:
+      not:
+        $ref: '#/definitions/rule'
+    required:
+      - not
+    type: object
+  oneOfRule:
+    additionalProperties: false
+    description: Asserts that exactly one of the nested rules matches.
+    properties:
+      oneOf:
+        $ref: '#/definitions/rules'
+    type: object
+  rule:
+    description: Defines single matching rule.
+    oneOf:
+      - $ref: '#/definitions/allOfRule'
+      - $ref: '#/definitions/anyOfRule'
+      - $ref: '#/definitions/matchRule'
+      - $ref: '#/definitions/notRule'
+      - $ref: '#/definitions/oneOfRule'
+  rules:
+    description: A list of matching rules.
+    items:
+      $ref: '#/definitions/rule'
+    type: array
+description: Schema of the sops-compliance-checker configuration file
+properties:
+  rules:
+    $ref: '#/definitions/rules'
+title: sops-compliance-checker configuration
+type: object
+```
+
+This is highly influenced by the [JSON Schema Specification][jsonschema-spec]
+itself to allow for great flexibility in the rule composition.
+
+### Configuration Example
+
+Given the following requirements:
+
+- The SOPS file must contain all necessary approved trust anchors:
+  - AGE recovery key
+    `age1u79ltfzz5k79ex4mpl3r76p2532xex4mpl3z7vttctudr6gedn6ex4mpl3` used for
+    disaster recovery.
+  - AWS KMS keys of **at least one team** owning the component. Additional team
+    keys are allowed, as long as they are part of the list of approved trust
+    anchors.
+  - AWS KMS deploy keys for **exactly one** target environment
+    (development/staging/production).
+- AWS KMS keys **must come in pairs** of two different regions to increase availability.
+- Any excess trust anchors not matching any of these rules **must be rejected**.
+
+One possible configuration to acheive this could be:
+
+```yaml
+---
+rules:
+  # All rules must match. This will automatically reject excess trust anchors
+  # not matching any of the nested rules.
+  - allOf:
+      # Disaster recovery key must be present.
+      - match: age1u79ltfzz5k79ex4mpl3r76p2532xex4mpl3z7vttctudr6gedn6ex4mpl3
```

Comment:
I'm wondering if we need to also specify the key type (AGE, PGP, AWS, GCP ...) or if the format is sufficiently different to not require it?

Reply:
From what I found so far, the key ID/fingerprint formats are unambiguous across all supported key types, so explicitly specifying the key format might not be necessary.

If this turns out to be not true, we could add a `type` field (or `format`) which is an enum over all key types SOPS supports (`kms`, `gcp_kms`, `azure_kv`, `hc_vault`, `age` and `pgp` as of now).